doc/design/csp.md

```diff
@@ -109,52 +109,43 @@ fmt.Println("Received from channel")
 
 In Fluid, we should be able to perform the above operations on the channel objects as well. As of now, we support two different kinds of channels : [Buffered Channel](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/details/buffered_channel.h) and [UnBuffered Channel](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/details/unbuffered_channel.h)
 
-Send and Receive can be performed as following on a buffered channel:
+With a buffered channel, the send operation blocks if the buffer is full, and the receive operation blocks if the buffer is empty.  The following program should run without blocking because the amount of data written into the buffer is less than the buffer size.
 
 ```python
-import threading
-
-def send_to_channel(channel, num_time=1):
-  for i in xrange(num_time):
-    channel.send(i)
+import fluid
 
 # Create a buffered channel of capacity 10
-buffer_size = 10;
+buffer_size = 10
 ch = fluid.make_channel(dtype=INT, buffer_size)
 
 # Now write three elements to the channel
-thread = threading.Thread(target=send_to_channel, args=(ch, 3, ))
-thread.daemon = True
-thread.start()
+with fluid.while(steps=buffer_size):
+  fluid.send(ch, step)
 
 # Read all the data from the channel
-for i in xrange(3):
-  y = ch.recv()
+with fluid.while(steps=buffer_size):
+  fluid.print(fluid.recv(ch))
 
 # Done receiving , now close the channel
-ch.close()
+fluid.close_channel(ch)
 ```
 
-The send and receive operations will be similar for unbuffered channel as well, except for the fact that there is no buffer in an unbuffered channel, so the operations are completely synchronized. For example:
+Both send and receive must block with a unbuffered channel, because the buffer size is 0.
 
 ```python
-import threading
-
-def send_to_channel(channel, data):
-  channel.send(data)
+import fluid
 
 # Create an unbuffered channel
 ch = fluid.make_channel(dtype=INT)
 
 # Writes and Reads are synchronous otherwise the calls will block.
-thread = threading.Thread(target=send_to_channel, args=(ch, 10, ))
-thread.daemon = True
-thread.start()
+with fluid.go():
+  fluid.send(ch)
 
-y = ch.recv()
+y = fluid.recv(ch)
 
 # Done receiving , now close the channel
-ch.close()
+fluid.close_channel(ch)
```

Comment:
Typically sender will close the channel, since send to a closed channel will panic, but receive from a closed channel will get zero value, optionally one can check if the value is valid:

```
a := make(chan int)
close(a)
fmt.Println(<-a) // 0
v, ok := <-a
fmt.Println(v, ok) // 0, false
```